src/main/java/com/conveyal/r5/analyst/cluster/GridResultAssembler.java

```diff
@@ -0,0 +1,201 @@
+ */
+public class GridResultAssembler {
+    public static final Logger LOG = LoggerFactory.getLogger(GridResultAssembler.class);
+
+    /** The version of the access grids we produce */
+    public static final int VERSION = 0;
+
+    /** The version of the origins we consume */
+    public static final int ORIGIN_VERSION = 0;
+
+    /** The offset to get to the data section of the access grid file */
+    public static final long DATA_OFFSET = 9 * 4;
+
+    private static final AmazonSQS sqs = new AmazonSQSClient();
+    private static final AmazonS3 s3 = new AmazonS3Client();
+
+    private Base64.Decoder base64 = Base64.getDecoder();
+
+    private final GridRequest request;
+
+    private File temporaryFile;
+    private RandomAccessFile buffer;
+
+    private boolean error = false;
+
+    /** this does not need to be an atomic int as it's incremented in a synchronized block */
+    public int nComplete = 0;
+
+    public int nTotal;
+
+    private final String outputBucket;
+
+    public GridResultAssembler (GridRequest request, String outputBucket) {
+        this.request = request;
+        this.outputBucket = outputBucket;
+        nTotal = request.width * request.height;
+    }
+
+    private void finish () {
+        // gzip and push up to S3
+        LOG.info("Finished receiving data for regional analysis {}, uploading to S3", request.jobId);
+
+        try {
+            File gzipFile = File.createTempFile(request.jobId, ".access_grid.gz");
+
+            buffer.close();
+
+            // There's probably a more elegant way to do this with NIO and without closing the buffer
+            InputStream is = new BufferedInputStream(new FileInputStream(temporaryFile));
+            OutputStream os = new GZIPOutputStream(new BufferedOutputStream(new FileOutputStream(gzipFile)));
+            ByteStreams.copy(is, os);
+            is.close();
+            os.close();
+
+            LOG.info("GZIP compression reduced regional analysis {} from {}mb to {}mb ({}x compression)",
+                    request.jobId,
+                    temporaryFile.length() / 1024 / 1024,
+                    gzipFile.length() / 1024 / 1024,
+                    temporaryFile.length() / gzipFile.length()
+                    );
+
+            s3.putObject(outputBucket, String.format("%s.access", request.jobId), gzipFile);
+
+            // these will fill up the disk lickety-split if we don't keep them cleaned up.
+            gzipFile.delete();
+            temporaryFile.delete();
+        } catch (Exception e) {
+            LOG.error("Error uploading results of regional analysis {}", request.jobId, e);
+        }
+    }
+
+    /** Process an SQS message */
+    public void handleMessage (Message message) {
+        try {
+            byte[] body = base64.decode(message.getBody());
+            ByteArrayInputStream bais = new ByteArrayInputStream(body);
```

Comment:
These variables need a naming scheme that is parallel with the origin output streams below, e.g. messageByteInputStream and messageDataInputStream. However even more advantageous would be to abstract the origin grids out into a separate class, which buffers the data for a single origin, runs header and sanity checks for that data, and has a method for writing out the delta-coded version to an arbitrary DataOutputStream. Or can write to an arbitrary region of a seekable buffer.